src/passes/Heap2Local.cpp

```diff
@@ -688,6 +707,27 @@ struct Struct2Local : PostWalker<Struct2Local> {
     replaceCurrent(builder.makeBlock(contents));
   }
 
+  void visitRefEq(RefEq* curr) {
+    if (!analyzer.reached.count(curr)) {
+      return;
+    }
+
+    if (curr->type == Type::unreachable) {
+      // The result does not matter. Leave things as they are (and let DCE
+      // handle it).
+      return;
+    }
+
+    // If our reference is compared to itself, the result is 1. If it is
+    // compared to something else, the result must be 0, as our reference does
+    // not escape to any other place.
+    int32_t result = analyzer.reached.count(curr->left) > 0 &&
+                     analyzer.reached.count(curr->right) > 0;
```

Comment:
How is this checking whether the expression is compared to itself?

Reply:
`reached` is the places the allocation reaches. So if the allocation is on both sides, we must be equal. If the allocation reached only one, then since it did not escape to a place we can't analyze, the other arm must be something totally different.

Reply:
What if one (or both) of the arms may or may not be the analyzed allocation?

Reply:
We are never in an ambiguous situation. If one arm can contain a mixture of the analyzed allocation and something else then we consider that the `Mixes` situation, which we ignore:

https://github.com/WebAssembly/binaryen/blob/60050be9abc2660a2a9cf97f9b83ccdf642b3f79/src/passes/Heap2Local.cpp#L243-L248

In theory we could add runtime checks "is this the allocation", but it [seems unpromising](https://github.com/WebAssembly/binaryen/blob/60050be9abc2660a2a9cf97f9b83ccdf642b3f79/src/passes/Heap2Local.cpp#L112-L150).